### PR TITLE
Improve AMP stability

### PR DESCRIPTION
**Pull request by michaelmckinsey1**
- [x] Compute dice loss in FP32 to avoid `val_dice_score=nan`
- [x] Use BF16 to prevent numerical overflow.
- [x] The datatype for AMP is now centrally configurable in `ScaFFold/utils/data_types.py` for testing`.

### CODE REVIEW

File: ScaFFold/utils/evaluate.py [hunk 1]
```diff
@@ -85,13 +90,23 @@ def foreground_dice_mean(dice_scores):
...
+                mask_true_onehot = (
+                    F.one_hot(local_labels, n_categories + 1)
+                    .permute(0, 4, 1, 2, 3)
+                    .float()
+                )
+
+                # Dice loss uses probabilities
+                dice_score_probs = compute_sharded_dice(
+                    mask_pred_probs, mask_true_onehot, spatial_mesh
+                )
```
Comment (PatrickRMiles) [inline ScaFFold/utils/evaluate.py]: It looks like this got inserted in the middle of the CE loss calc. Can you move it back to being after `CE_loss = ...`? This should also shrink the diff + make it more clear what the actual changes are here (not casting local_preds to float)

File: ScaFFold/utils/trainer.py [hunk 2]
```diff
@@ -194,7 +201,8 @@ def setup_training_components(self):
         )
 
         # Set up gradient scaler for AMP (Automatic Mixed Precision)
-        self.grad_scaler = torch.amp.GradScaler("cuda", enabled=self.config.torch_amp)
+        self.use_grad_scaler = self.config.torch_amp and self.amp_dtype == torch.float16
```
Comment (PatrickRMiles) [inline ScaFFold/utils/trainer.py]: Is the `and self.amp_dtype == torch.float16` basically just catching the case where we're NOT running with bf16? Would it be better to write that explicitly, like `and self.amp_dtype != torch.bfloat16`?

File: ScaFFold/utils/trainer.py [hunk 3]
```diff
@@ -428,14 +442,23 @@ def warmup(self):
...
+                    local_labels_one_hot = (
+                        F.one_hot(
+                            local_labels, num_classes=self.config.n_categories + 1
+                        )
+                        .permute(0, 4, 1, 2, 3)
+                        .float()
+                    )
+                    dice_scores = compute_sharded_dice(
+                        local_preds_softmax, local_labels_one_hot, self.spatial_mesh
+                    )
```
Comment (PatrickRMiles) [inline ScaFFold/utils/trainer.py]: Same as in `evaluate.py`, I think this should stay after `global_ce_sum = ...`

File: ScaFFold/utils/trainer.py [hunk 4]
```diff
@@ -204,15 +212,24 @@ def setup_training_components(self):
         )
 
         self.log.info(
-            f"Optimizer: {self.optimizer}, Scheduler: {self.scheduler}, Gradient Scaler Enabled: {self.config.torch_amp}"
+            f"Optimizer: {self.optimizer}, Scheduler: {self.scheduler}, AMP dtype: {self.amp_dtype}, Gradient Scaler Enabled: {self.use_grad_scaler}"
         )
 
+    def _autocast_kwargs(self, enabled=None):
```
Comment (PatrickRMiles) [inline ScaFFold/utils/trainer.py:221]: Just so I understand, the purpose of the `enabled` arg is that it allows us to force autocast on/off regardless of `self.config.torch_amp`?

File: ScaFFold/utils/trainer.py [hunk 5]
```diff
@@ -627,18 +638,31 @@ def train(self):
...
+                                    )
+                                    .permute(0, 4, 1, 2, 3)
+                                    .float()
+                                )
+                                # Compute sharded dice
+                                dice_scores = compute_sharded_dice(
+                                    local_preds_softmax,
+                                    local_labels_one_hot,
+                                    self.spatial_mesh,
+                                )
```
Comment (PatrickRMiles) [inline ScaFFold/utils/trainer.py]: Same as evaluate and warmup, this should come after `global_ce_sum = ...`

**Review by PatrickRMiles — CHANGES_REQUESTED** [review]

File: ScaFFold/utils/trainer.py (hunk 2 above)
Comment (michaelmckinsey1) [reply ScaFFold/utils/trainer.py]: Yes basically the options are `bf16` or `f16`. I don't think `bf/f8` would work. I am ok with making this change

File: ScaFFold/utils/trainer.py (hunk 4 above)
Comment (michaelmckinsey1) [reply ScaFFold/utils/trainer.py:221]: Its to parametrize regions where we want dtype to be `float32`, otherwise dtype will be `amp_dtype`. It could also be `enabled=True` by default and then explicitly `enabled=False` in float32 regions

File: ScaFFold/utils/evaluate.py (hunk 1 above)
Comment (michaelmckinsey1) [reply ScaFFold/utils/evaluate.py]: done

File: ScaFFold/utils/trainer.py (hunk 3 above)
Comment (michaelmckinsey1) [reply ScaFFold/utils/trainer.py]: done

File: ScaFFold/utils/trainer.py (hunk 5 above)
Comment (michaelmckinsey1) [reply ScaFFold/utils/trainer.py]: done

**Review by PatrickRMiles — APPROVED** [review]
